usr/src/uts/armv8/os/mp_startup.c

```diff
@@ -710,31 +690,48 @@ start_other_cpus(int cprboot)
 	affinity_set(CPU_CURRENT);
 
 	mutex_enter(&cpu_lock);
-	for (int who = 0; who < NCPU; who++) {
-		if (!CPU_IN_SET(mp_cpus, who))
+
+	/*
+	 * Set up CPU structures for each CPU we're going to start.
+	 */
+	for (struct cpuinfo *ci = cpuinfo_first_enabled();
+	    ci != cpuinfo_end(); ci = cpuinfo_next_enabled(ci)) {
+		if (ci->ci_id == bootcpu || !CPU_IN_SET(mp_cpus, ci->ci_id))
 			continue;
-		cpu_t *cp;
-		cp = mp_cpu_configure_common(who, CPU->cpu_m.affinity);
-		ASSERT(cp != NULL);
+		mp_cpu_configure_common(ci);
 	}
-	for (int who = 0; who < NCPU; who++) {
-		if (!CPU_IN_SET(mp_cpus, who))
-			continue;
-		ASSERT(who != bootcpuid);
 
-		if (start_cpu(who, CPU->cpu_m.affinity) != 0)
-			CPUSET_DEL(mp_cpus, who);
-		cpu_state_change_notify(who, CPU_SETUP);
+	/*
+	 * Start any enabled non-boot CPUs.
+	 */
+	for (struct cpuinfo *ci = cpuinfo_first_enabled();
+	    ci != cpuinfo_end(); ci = cpuinfo_next_enabled(ci)) {
+		if (ci->ci_id == bootcpu || !CPU_IN_SET(mp_cpus, ci->ci_id))
+			continue;
+		/*
+		 * XXXARM: The error path here needs to be tightened up, as
+		 * when this fails we will crash shortly thereafter.
+		 */
+		if (start_cpu(ci) != 0)
+			CPUSET_DEL(mp_cpus, ci->ci_id);
+		cpu_state_change_notify(ci->ci_id, CPU_SETUP);
 		mutex_exit(&cpu_lock);
 		mutex_enter(&cpu_lock);
 	}
-	for (int who = 0; who < NCPU; who++) {
-		if (!CPU_IN_SET(mp_cpus, who))
-			continue;
-		cpu_t *cp = cpu[who];
-	}
 
 	mutex_exit(&cpu_lock);
 
 	affinity_clear();
+
+	cmn_err(CE_CONT,
```

Comment:
Is this debug output from your work, or something we want to leave in?

Reply:
Yeah, I intended to remove this before submitting the PR, but missed it - will clean up.